### PR TITLE
Feature/pm 194 updating gait selection tests

### PR DESCRIPTION
**Pull request by JorisWeeda**
Closes PM-194

## Description
Updated the unit-test of the gait selection module to verify every class function. Divided the gait selection launch file and the dynamic gaits launch in order to set the balance argument. Within this new structure it is possible to write tests for the move-it related functions.  

## Changes
* Updated the unit test of the gait selection module
* Added some additional checks in the gait selection class
* Created separate launch file for the dynamic gait tests
* Created new testing file for the balance gait

### CODE REVIEW

File: march_gait_selection/test/nosetests/test_gait_balance.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    def setUpClass(cls):
+        cls._gait_selection = GaitSelection(PKG, DIR)
+
+    def setUp(self):
+        self.gait_selection = deepcopy(self._gait_selection)
+
+    # __init__ tests
+    def test_get_balance_walk_gait(self):
+        is_balanced_walk = True if self.gait_selection['gait_balanced_walk'] is not None else False
+        self.assertTrue(is_balanced_walk)
```
Comment (RutgerVanBeek) [inline march_gait_selection/test/nosetests/test_gait_balance.py:25]:
```suggestion
        self.assertFalse(self.gait_selection['gait_balanced_walk'] is None)
```

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 2]
```diff
@@ -1,78 +1,90 @@
...
-
-right_open = 'right_open'
-right_swing = 'right_swing'
-right_close = 'right_close'
-
-wrong_name = 'wrong'
-
-gait_selection = GaitSelection(PKG, DIR)
+valid_package = 'march_gait_selection'
+valid_directory = 'test/testing_gait_files'
```
Comment (RutgerVanBeek) [inline march_gait_selection/test/nosetests/test_gait_selection.py:13]:
```suggestion
VALID_PACKAGE = 'march_gait_selection'
VALID_DIRECTORY = 'test/testing_gait_files'
```

File: march_gait_selection/test/nosetests/test_gait_balance.py (hunk 1 above)
Comment (RutgerVanBeek) [inline march_gait_selection/test/nosetests/test_gait_balance.py:25]:
```suggestion
        self.assertFalse(self.gait_selection['gait_balanced_walk'] is None)
```

File: march_gait_selection/test/nosetests/test_gait_selection.py (hunk 2 above)
Comment (RutgerVanBeek) [inline march_gait_selection/test/nosetests/test_gait_selection.py:13]:
```suggestion
VALID_PACKAGE = 'march_gait_selection'
VALID_DIRECTORY = 'test/testing_gait_files'
```

File: march_gait_selection/test/nosetests/test_gait_balance.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    def test_get_balance_walk_gait_type(self):
+        balanced_walk = self.gait_selection['gait_balanced_walk']
+        self.assertIsInstance(balanced_walk, BalanceGait)
+
+    def test_initialised_move_groups(self):
+        balanced_walk = self.gait_selection['gait_balanced_walk']
+        has_move_groups = True if balanced_walk._move_groups is not None else False
+        self.assertTrue(has_move_groups)
```
Comment (RutgerVanBeek) [inline march_gait_selection/test/nosetests/test_gait_balance.py:34]:
```suggestion
        has_move_groups = True if balanced_walk._move_groups is not None else False
        self.assertFalse(balanced_walk._move_groups is None)
```

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 4]
```diff
@@ -1,78 +1,90 @@
...
+        self.assertEqual(walk_subgaits['left_swing'], ['MV_walk_leftswing_v2'])
+
+    # get item tests
+    def test_get_item_with_wrong_name(self):
+        gait = self.gait_selection['wrong']
+        self.assertEqual(gait, None)
+
+    def test_get_item_valid_gait_name(self):
+        gait = True if self.gait_selection['walk'] is not None else False
+        self.assertTrue(gait)
```
Comment (RutgerVanBeek) [inline march_gait_selection/test/nosetests/test_gait_selection.py:86]:
```suggestion
        gait = True if  else False
        self.assertFalse(self.gait_selection['walk'] is None)
```

File: march_gait_selection/src/march_gait_selection/gait_selection.py [hunk 5]
```diff
@@ -105,6 +111,9 @@ def scan_directory(self):
     def validate_versions_in_directory(self, new_gait_version_map):
         """Validate if the given version numbers in the version map exist in the selected directory."""
         for gait_name in new_gait_version_map:
+            if not self.validate_gait_in_directory(gait_name):
+                return False
```
[the remark this answers is not in the document]
Comment (JorisWeeda) [reply march_gait_selection/src/march_gait_selection/gait_selection.py:115]: indeed!

File: march_gait_selection/test/nosetests/test_gait_balance.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import unittest
+
+from march_gait_selection.dynamic_gaits.balance_gait import BalanceGait
+from march_gait_selection.gait_selection import GaitSelection
+
+PKG = 'march_gait_selection'
+DIR = 'test/testing_gait_files'
+
+
+class TestTransitionTrajectory(unittest.TestCase):
```
Comment (Olavhaasie) [inline march_gait_selection/test/nosetests/test_gait_balance.py:13]: This test case has the same name as the one in `test_gait_transition.py` and you don't seem to run it anywhere. It probably runs because it has the same name (???). I think it would be best to give it a unique name and it to the `unitrun`.

Comment (JorisWeeda) [reply march_gait_selection/test/nosetests/test_gait_balance.py:13]: ooopsie ):

**Comment by JorisWeeda** [comment]
> Can you also include a test where you actually try to plan something with the move_groups (may be use random_subgait for this) to see if the move_groups have initialized well. If you think this should be done after we actually walked with balanced gaits then that is also okay. (May be make an issue in Balance JIRA to create tests then).

My initial thought was to create a separate issue for the balance tests. This should only provide a working test suite to easy the adding in the future. I will create a new JIRA issue.

> Furthermore it might be nice to also test that a succesfully loaded gait is not only a gait, but also returns subgaits. (Or is this more something to test in the march_shared_classes?)

If the gait object is passed correctly using the get item method the subgait should be available as well. This is indeed checked in the test classes of the _march_shared_classes_.

File: march_gait_selection/src/march_gait_selection/gait_selection.py [hunk 7]
```diff
@@ -138,3 +145,7 @@ def update_default_versions(self):
 
         except IOError:
             return [False, 'Error occurred when writing to file path: {pn}'.format(pn=self.default_yaml)]
+
+    def __getitem__(self, name):
```
[the remark this answers is not in the document]
Comment (JorisWeeda) [reply march_gait_selection/src/march_gait_selection/gait_selection.py:149]: I could add some more dictionary like functionalities but they are not used at any point in code, would be nice to have but not really necessary in my opinion.

File: march_gait_selection/test/nosetests/test_gait_balance.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    @classmethod
+    def setUpClass(cls):
+        cls._gait_selection = GaitSelection(VALID_PACKAGE, VALID_DIRECTORY)
+
+    def setUp(self):
+        self.gait_selection = deepcopy(self._gait_selection)
+
+    # __init__ tests
+    def test_get_balance_walk_gait(self):
+        self.assertFalse(self.gait_selection['gait_balanced_walk'] is None)
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_balance.py:24]:
```suggestion
        self.assertIsNotNone(self.gait_selection['gait_balanced_walk'])
```

File: march_gait_selection/test/nosetests/test_gait_balance.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    def test_get_balance_walk_gait_type(self):
+        balanced_walk = self.gait_selection['gait_balanced_walk']
+        self.assertIsInstance(balanced_walk, BalanceGait)
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_balance.py:28]:
```suggestion
        self.assertIsInstance(self.gait_selection['gait_balanced_walk'], BalanceGait)
```

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 10]
```diff
@@ -1,78 +1,89 @@
...
+
+    # load gaits tests
+    def test_types_in_loaded_gaits(self):
+        for gait in self.gait_selection.loaded_gaits:
+            self.assertIsInstance(gait, (Gait, BalanceGait))
+
+    # scan directory tests
+    def test_scan_directory_top_level_length(self):
+        directory = self.gait_selection.scan_directory()
+        self.assertTrue(len(['stairs_up', 'walk', 'walk_medium', 'walk_small']), len(directory.keys()))
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_selection.py:68]:
```suggestion
        self.assertEqual(len(['stairs_up', 'walk', 'walk_medium', 'walk_small']), len(directory.keys()))
```
What do you want to test here? why not test if len(directory.keys()) == 4 or
directory.keys() == ['stairs_up', 'walk', 'walk_medium', 'walk_small']

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 11]
```diff
@@ -1,78 +1,89 @@
...
+
+    def test_scan_directory_top_level_content(self):
+        directory = self.gait_selection.scan_directory()
+        self.assertTrue(all(key in ['stairs_up', 'walk', 'walk_medium', 'walk_small'] for key in directory.keys()))
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_selection.py:72]:
```suggestion
        self.assertEqual( ['stairs_up', 'walk', 'walk_medium', 'walk_small'], directory.keys() )
```
I believe you can do this

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 12]
```diff
@@ -1,78 +1,89 @@
...
+
+    def test_scan_directory_subgait_versions(self):
+        directory = self.gait_selection.scan_directory()
+        walk_subgaits = directory['walk']['subgaits']
+        self.assertEqual(walk_subgaits['left_swing'], ['MV_walk_leftswing_v2'])
+
+    # get item tests
+    def test_get_item_with_wrong_name(self):
+        gait = self.gait_selection['wrong']
+        self.assertEqual(gait, None)
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_selection.py:82]:
```suggestion
        self.assertIsNone(self.gait_selection['wrong'])
```

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 13]
```diff
@@ -1,78 +1,89 @@
...
+
+    def test_get_item_valid_gait_name(self):
+        self.assertFalse(self.gait_selection['walk'] is None)
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_selection.py:85]:
```suggestion
        self.assertIsNotNone(self.gait_selection['walk'])
```

File: march_gait_selection/test/nosetests/test_gait_selection.py [hunk 14]
```diff
@@ -1,78 +1,89 @@
...
+
+    def test_get_item_type(self):
+        gait = self.gait_selection['walk']
+        self.assertIsInstance(gait, Gait)
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_selection.py:89]:
```suggestion
        self.assertIsInstance(self.gait_selection['walk'], Gait)
```
(doing this makes the error message a bit more informative)

File: march_gait_selection/test/nosetests/test_gait_selection.py (hunk 11 above)
Comment (JorisWeeda) [reply march_gait_selection/test/nosetests/test_gait_selection.py:72]: I could do this but I have to change the order of the first list, In my opinion this solution is better.

File: march_gait_selection/test/nosetests/test_gait_selection.py (hunk 10 above)
Comment (JorisWeeda) [reply march_gait_selection/test/nosetests/test_gait_selection.py:68]: If all the maps included in the test resources are parsed properly!

File: march_gait_selection/test/nosetests/test_gait_balance.py [hunk 15]
```diff
@@ -14,15 +14,18 @@ class TestBalanceGait(unittest.TestCase):
 
     @classmethod
     def setUpClass(cls):
-        cls._gait_selection = GaitSelection(VALID_PACKAGE, VALID_DIRECTORY)
-
-    def setUp(self):
-        self.gait_selection = deepcopy(self._gait_selection)
```
Comment (Roelemans) [inline march_gait_selection/test/nosetests/test_gait_balance.py]: Why did you remove this?

Comment (JorisWeeda) [reply march_gait_selection/test/nosetests/test_gait_balance.py]: copying the balanceGait attribute from the gait_selection seems to be an issue when testing with _colcon test_. My guess is that it has something to do with the moveit related initiated values.

Comment (JorisWeeda) [reply march_gait_selection/test/nosetests/test_gait_balance.py]: Sadly this isn't the only problem, the moveit launch is also rejected/not-foud when testing in travis CI ):

Comment (Roelemans) [reply march_gait_selection/test/nosetests/test_gait_balance.py]: I see, that's unfortunate. In that case we need to pay attention that the tests do not modify the self.gait_selection object as these changes will remain in other tests
